pkg/syncer/storage_policy_quota_reconciler.go

```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		log.Errorf("newStoragePolicyQuotaReconciler: could not retrieve dynamic informer factory. Error: %+v", err)
+		return nil, err
+	}
+	// Return informer from shared dynamic informer factory for input resource.
+	gvr := schema.GroupVersionResource{Group: cnsoperatorv1alpha1.GroupName, Version: cnsoperatorv1alpha1.Version,
+		Resource: cnsoperatorv1alpha1.CnsStoragePolicyQuotaPlural}
+	storagePolicyQuotaInformer := dynamicInformerFactory.ForResource(gvr)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to get informer with error: %v", err)
+	}
+	svcQuotaOpsQueue := workqueue.NewNamedRateLimitingQueue(
+		svcStoragePolicyQuotaRateLimiter, "storage-policy-quota-ops")
+
+	rc := &storagePolicyQuotaReconciler{
+		metadataSyncerInformer: metadataSyncerInformer,
+		svcQuotaSynced:         storagePolicyQuotaInformer.Informer().HasSynced,
+		svcQuotaOpsQueue:       svcQuotaOpsQueue,
+	}
+
+	_, err = storagePolicyQuotaInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    rc.addStoragePolicyQuota,
+			DeleteFunc: rc.delStoragePolicyQuota,
+		})
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to add event handler on "+
+			"StoragePolicyQuota informer. Error: %v", err)
+	}
+
+	// Start StoragePolicyQuota Informer.
+	dynamicInformerFactory.Start(stopCh)
+	if !cache.WaitForCacheSync(stopCh, rc.svcQuotaSynced) {
+		return nil, fmt.Errorf("newStoragePolicyQuotaReconciler: cannot sync StoragePolicyQuota cache")
+	}
+
+	return rc, nil
+}
+
+func (rc *storagePolicyQuotaReconciler) addStoragePolicyQuota(obj interface{}) {
+	_, log := logger.GetNewContextWithLogger()
+	policyQuotaObj := &storagepolicyusagev1alpha1.StoragePolicyQuota{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object,
+		policyQuotaObj)
+	if err != nil {
+		log.Errorf("addStoragePolicyQuota: failed to cast object %+v to %s. Error: %v", obj,
+			cnsoperatorv1alpha1.CnsStoragePolicyQuotaSingular, err)
+		return
+	}
+	rc.svcQuotaOpsQueue.Add(policyQuotaObj)
+}
+
+func (rc *storagePolicyQuotaReconciler) delStoragePolicyQuota(obj interface{}) {
+	_, log := logger.GetNewContextWithLogger()
+	policyQuotaObj := &storagepolicyusagev1alpha1.StoragePolicyQuota{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object,
+		policyQuotaObj)
+	if err != nil {
+		log.Errorf("delStoragePolicyQuota: failed to cast object %+v to %s. Error: %v", obj,
+			cnsoperatorv1alpha1.CnsStoragePolicyQuotaSingular, err)
+		return
+	}
+	// Add annotation locally on StoragePolicyQuota object retrieved before adding it to queue,
+	// to identify that object being deleted during its processing by worker threads.
+	policyQuotaObj.Annotations[deletionAnnKey] = deletionAnnVal
+
+	rc.svcQuotaOpsQueue.Add(policyQuotaObj)
```

Comment:
If there is an issue where the deletion timestamp is not set, then create a struct that has {spq, isDelete}, and add that item into the queue. When you get from a queue, cast it to that struct, if isDelete=true process it further to remove the SPU.

Reply:
added new struct with StoragePolicyQuota objct pointer and bool to indicate if deletion or not